pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -371,6 +371,10 @@ func (m swapConfigurationHelper) ConfigureLimitedSwap(lcr *runtimeapi.LinuxConta
 }
 
 func (m swapConfigurationHelper) ConfigureNoSwap(lcr *runtimeapi.LinuxContainerResources) {
+	if !swapControllerAvailable() {
```

Comment:
As a note to the reviewers: It may be better to include this check in other code paths such as `ConfigureUnlimitedSwap`. Initially I was thinking that it would be better to do it in `configureSwap`, but would have to then handle the error cases where we are expecting to set up swap.

Reply:
How about moving it to here?

https://github.com/kubernetes/kubernetes/blob/f1703405243e6e6332d4b1a68d65bfcdad66bc05/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L110-L110


Then, if the feature gate or the cgroup controller are missing, swap logic wouldn't be executed at all.
WDYT?

Reply:
Yes, I was mulling over that last night too. That might be the simplest solution for now.

Reply:
> This should only be an issue if LimitedSwap is configured though -- since on a system with swap disabled my current understanding is that max and 0 are equivilent in terms of how these are handled by both Containerd and runc.

If swap is disabled, I believe that `0`, `max`, or any other value would serve as a no-op. However, in the case where the swap controller is not present, the `memory.swap.max` file is not present and therefore cannot be written to.

> One advantage of changing the swapConfigurationHelper implementation is that it allows for errors to be logged when a requested configuration cannot be applied.

If the swap controller is not available we can't use any of swapConfigurationHelper's methods, since they all configure `memory.swap.max` which is unavailable. This is why if you implement it on the swapConfigurationHelper level you would need to check for the controller's availability within each and every one of its methods, and then handle logging and returning as a no-op in each method separately.

Since the helper is being used only in one single place, I think that checking for the controller's availability in the same place is both more clear and less error-prone, as if another method would be added in the future it won't need to explicitly check if the controller is available once again. As I see it, when the helper is being used it already assumed that swap is enabled.

Does that make sense to you?

Reply:
Yes, I think that makes sense. I also played around with the helper a bit last night and defined an interface:
```
type swapConfigurer interface {
   configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64)
}
```
With three different implementations:
* `disabledSwapConfigurer`
* `unifiedSwapConfigurer`
* `nonUnifiedSwapConfigurer`

```
func newSwapConfigurer() swapConfigurer {
	if !swapControllerAvailable() {
		return &swapDisabledConfigurer{}
	}
	if !isCgroup2UnifiedMode() {
		return &nonUnifiedSwapConfigurer{}
	}
	return &unifiedSwapConfigurer{}
}

type swapConfigurer interface {
	configureSwap(*runtimeapi.LinuxContainerResources, int64)
}

// swapDisabledConfigurer represents a swap configurer where any operation to set the swap is a no-op.
// In cases where swap is set to a non-zero value, an error is logged.
type swapDisabledConfigurer struct{}

func (m swapDisabledConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if swapMemory == 0 {
		return
	}
	klog.ErrorS(fmt.Errorf("swap configuration is not supported on this system"), "swap configuration is unexpected if swap is disabled")
}

// unifiedSwapConfigurer represents a swap configurer where the swap is set in the unified cgroup v2 hierarchy.
// This sets the `memory.swap.max` cgroup value.
type unifiedSwapConfigurer struct{}

func (m unifiedSwapConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if lcr.Unified == nil {
		lcr.Unified = map[string]string{}
	}

	var swapString string
	if swapMemory == unlimitedSwap {
		swapString = "max"
	} else {
		swapString = fmt.Sprintf("%d", swapMemory)
	}
	lcr.Unified[cm.Cgroup2MaxSwapFilename] = swapString
}

// nonUnifiedSwapConfigurer represents a swap configurer where the swap is set in the non-unified cgroup v1 hierarchy.
// Unless swap is being disabled (or set ot max), an error is logged.
type nonUnifiedSwapConfigurer struct{}

func (m nonUnifiedSwapConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if swapMemory == 0 || swapMemory == unlimitedSwap {
		// memorySwapLimit = total permitted memory+swap; if equal to memory limit, => 0 swap above memory limit
		// Some swapping is still possible.
		// Note that if memory limit is 0, memory swap limit is ignored.
		lcr.MemorySwapLimitInBytes = lcr.MemoryLimitInBytes
	}
	klog.ErrorS(fmt.Errorf("swap configuration is not supported with cgroup v1"), "swap configuration under cgroup v1 is unexpected")
}
```

I updated the helper as follows:
```
type swapConfigurationHelper struct {
	swapConfigurer
	machineInfo cadvisorv1.MachineInfo
}

func newSwapConfigurationHelper(machineInfo cadvisorv1.MachineInfo) *swapConfigurationHelper {
	configurer := newSwapConfigurer()
	return &swapConfigurationHelper{
		swapConfigurer: configurer,
		machineInfo:    machineInfo,
	}
}
```

meaning that it delegates to the swapConfigurer to accutally apply the swap value. I did feel that this was a larger change than what is required for this fix -- but if you think it's something worth following-up on, I don't mind creating a PR with the refactor to clarify things.

Reply:
Thanks @elezar! I appreciate the work you put into it.

My initial thoughts here is that 2 of the 3 implementations for your interface can actually be used only to configure no swap access or act as a no-op. In other words, only `unifiedSwapConfigurer` has real logic into it while `disabledSwapConfigurer` and `nonUnifiedSwapConfigurer` are almost identical in the sense that they either don't do anything or configure no swap.

In the current implementation, this is being solved by two `if` statements. The one checks if the controller is available and if it isn't the swapConfigHelper isn't even being used. The other one resides within `ConfigureNoSwap` and will likely be removed in the future when cgroup v1 would be fully dropped. Moreover, these two implementations aren't going to expend in the future.

To be completely honest, I think that your current implementation looks great. It is elegant and simple to understand. My opinion is that we don't need to complicate stuff before we have a good reason to do so (see [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)).

Perhaps we can circle back to it in the future if we indeed have more cases to handle